src/concurrent.rs

```diff
@@ -103,6 +113,145 @@ impl io::Read for ShareableFile {
+/// `SignalsHandler` object: the caller must call `prepare()` before mounting the file system and
+/// then call `install()` once the file system is ready to serve.
+///
+/// Keeping this logic as a separate `SignalsInstaller` object, instead of trying to expose a "safe
+/// mount function" helps ensure restoration of the original signal mask in all cases because this
+/// type does so at drop time.
+pub struct SignalsInstaller {
+    /// Signal mask to restore at drop time.
+    old_sigset: signal::SigSet,
+}
+
+impl SignalsInstaller {
+    /// Blocks signals in preparation to mount the file system.
+    pub fn prepare() -> SignalsInstaller {
+        let mut old_sigset = signal::SigSet::empty();
+        let mut sigset = signal::SigSet::empty();
+        for signal in CAPTURED_SIGNALS.iter() {
+            sigset.add(*signal);
+        }
+        signal::pthread_sigmask(
+            signal::SigmaskHow::SIG_BLOCK, Some(&sigset), Some(&mut old_sigset))
+            .expect("pthread_sigmask is not expected to fail");
+        SignalsInstaller { old_sigset }
+    }
+
+    /// Installs all signal handlers to unmount the given `mount_point`.
+    pub fn install(self, mount_point: PathBuf) -> Result<SignalsHandler, Error> {
+        let (signal_sender, signal_receiver) = mpsc::channel();
+
+        let mut signums = vec!();
+        for signal in CAPTURED_SIGNALS.iter() {
+            signums.push(*signal as i32);
+        }
+        let signals = signal_hook::iterator::Signals::new(&signums)?;
+
+        std::thread::spawn(move || SignalsHandler::handler(&signals, mount_point, &signal_sender));
+
+        Ok(SignalsHandler { signal_receiver })
+
+        // Consumes self which causes the original signal mask to be restored and thus unblocks
+        // signals.
+    }
+}
+
+impl Drop for SignalsInstaller {
+    fn drop(&mut self) -> () {
+        signal::pthread_sigmask(signal::SigmaskHow::SIG_SETMASK, Some(&self.old_sigset), None)
+            .expect("pthread_sigmask is not expected to fail and we cannot correctly clean up");
+    }
+}
+
+/// Tries to unmount the given file system indefinitely.
+///
+/// If unmounting fails, it is probably because the file system is busy.  We don't know but it
+/// doesn't matter: we have entered a terminal status: we do this at exit time so we'll keep trying
+/// to unclog things while telling the user what's going on.  They are the ones that have to fix
+/// this situation.
+fn retry_unmount<P: AsRef<Path>>(mount_point: P) {
+    let mut backoff = time::Duration::from_millis(10);
+    'retry: loop {
+        match fuse::unmount(mount_point.as_ref()) {
+            Ok(()) => break 'retry,
+            Err(e) => {
+                warn!("Unmounting file system failed with error '{}'; will retry in {:?}",
+                    e, backoff);
+                thread::sleep(backoff);
+                if backoff.as_secs() < 1 {
```

Comment:
If you want to cap at 1s and not 1.28s, you need a min function here.

Reply:
Done.